lua/telescope/builtin/lsp.lua

```diff
@@ -259,44 +259,43 @@ lsp.range_code_actions = function(opts)
 end
 
 lsp.workspace_symbols = function(opts)
-  local params = { query = opts.query or "" }
-  local results_lsp, err = vim.lsp.buf_request_sync(0, "workspace/symbol", params, opts.timeout or 10000)
-  if err then
-    vim.api.nvim_err_writeln("Error when finding workspace symbols: " .. err)
-    return
-  end
-
-  local locations = {}
-
-  if results_lsp and not vim.tbl_isempty(results_lsp) then
-    for _, server_results in pairs(results_lsp) do
-      -- Some LSPs (like Clangd and intelephense) might return { { result = {} } }, so make sure we have result
-      if server_results and server_results.result and not vim.tbl_isempty(server_results.result) then
-        vim.list_extend(locations, vim.lsp.util.symbols_to_items(server_results.result, 0) or {})
+  local function get_workspace_symbols_requester()
+    local cancel = function() end
+    local bufnr = vim.api.nvim_get_current_buf()
+    local params = { query = opts.query or "" }
+
+    return function()
+      local tx, rx = channel.oneshot()
+      cancel()
+      _, cancel = vim.lsp.buf_request(bufnr, "workspace/symbol", params, tx)
+
+      local err, _, results_lsp = rx()
+      assert(not err, err)
+
+      local locations = vim.lsp.util.symbols_to_items(results_lsp or {}, bufnr) or {}
+      locations = utils.filter_symbols(locations, opts)
+      if locations == nil then
+        -- error message already printed in `utils.filter_symbols`
+        return
       end
-    end
-  end
 
-  locations = utils.filter_symbols(locations, opts)
-  if locations == nil then
-    -- error message already printed in `utils.filter_symbols`
-    return
-  end
+      if vim.tbl_isempty(locations) then
+        print(
+          "No results from workspace/symbol. Maybe try a different query: "
+            .. "Telescope lsp_workspace_symbols query=example"
+        )
+        return
+      end
 
-  if vim.tbl_isempty(locations) then
-    print(
-      "No results from workspace/symbol. Maybe try a different query: "
-        .. "Telescope lsp_workspace_symbols query=example"
-    )
-    return
+      return locations
```

Comment:
What we can also say that we can return is a Iterator rather than a table. That way we could return a iterator that iterates over a tag file rather than that we need to split the tag file and return a table. Open for suggestions